File: src/engine/adapters/mysql-family/read.ts
import {
  type ColumnDescriptor,
  createTabularPageBuilder,
  type TabularPage,
  type TypeClass,
} from '@shared/protocol/page';
import type { Connection } from 'mariadb';
import type { OpCtx, ReadRequest } from '../adapter';
import { AdapterError } from '../errors';
import {
  assertKeysetSupported,
  buildKeysetPosition,
  buildKeysetPredicate,
  buildScanOrderBy,
  computeEffectiveOrder,
  decodePageToken,
  parseCountValue,
  requestFingerprint,
  resolveFetchColumns,
  resolveProjection,
  safeInt,
  whereClause,
} from '../sql-text';
import type { ReadTarget } from './catalog';
import { runQuery, type TrackQuery } from './query';

export function quoteIdent(name: string): string {
  if (name.includes('\0')) throw new AdapterError('E_QUERY', 'identifier contains a NUL byte');
  return `\`${name.replace(/`/g, '``')}\``;
}

// §5d's MariaDB mapping. tinyint(1) is checked ahead of the general number match — it is how
// MariaDB spells boolean.
export function typeClassFor(dataType: string): TypeClass {
  const base = dataType.toLowerCase();
  if (/^tinyint\(1\)/.test(base)) return 'boolean';
  if (
    /^(tinyint|smallint|mediumint|int|integer|bigint|decimal|numeric|float|double|bit)\b/.test(base)
  ) {
    return 'number';
  }
  if (/^(date|datetime|timestamp|time|year)\b/.test(base)) return 'temporal';
  if (base.startsWith('json')) return 'json';
  if (/^(binary|varbinary|tinyblob|blob|mediumblob|longblob|geometry)\b/.test(base))
    return 'binary';
  return 'text';
}

export async function readPage(
  conn: Connection,
  ctx: OpCtx,
  track: TrackQuery,
  target: ReadTarget,
  req: Omit<ReadRequest, 'path'>,
): Promise<TabularPage> {
  const projectedColumns = resolveProjection(target.columns, req.projection);
  const order = computeEffectiveOrder(
    req.sort,
    target.columns,
    target.primaryKey ?? target.uniqueKeys[0] ?? null,
  );
  const isTextSort = req.sort?.kind === 'text';
  const wantsKeyset = req.cursor.mode === 'after' || req.cursor.mode === 'before';
  assertKeysetSupported(wantsKeyset, isTextSort, order.keysetEligible);

  const { fetchColumns, keysetColumnIndex } = resolveFetchColumns(
    projectedColumns,
    target.columns,
    order,
  );

  const columns: ColumnDescriptor[] = projectedColumns.map((c) => ({
    name: c.name,
    dataType: c.dataType,
    typeClass: typeClassFor(c.dataType),
    nullable: c.nullable,
    isPrimaryKey: c.isPrimaryKey,
    // P36 D28: not detected here yet — false rather than a guess.
    generated: false,
  }));

  const relationSql = `${quoteIdent(target.qualifiedName.database)}.${quoteIdent(target.qualifiedName.table)}`;
  const selectList = fetchColumns.map((c) => quoteIdent(c.name)).join(', ');

  const params: unknown[] = [];
  let whereSql = whereClause(req.filter);

  const fingerprint = requestFingerprint({
    path: target.qualifiedName,
    projection: req.projection,
    filter: req.filter,
    sort: req.sort,
    pageSize: req.pageSize,
  });

  const reverseRows = req.cursor.mode === 'before' && order.keysetEligible;
  const orderBySql = buildScanOrderBy(req.sort, order, reverseRows, quoteIdent);

  if (wantsKeyset && req.cursor.mode !== 'offset') {
    const token = req.cursor.token;
    const keyValues = decodePageToken(token, fingerprint);
    if (keyValues.length !== order.keysetColumns.length) {
      throw new AdapterError('E_QUERY', 'page token key length does not match the sort key');
    }
    const quotedKeyColumns = order.keysetColumns.map((c) => quoteIdent(c));
    const predicate = buildKeysetPredicate(
      quotedKeyColumns,
      order.keysetDirection,
      req.cursor.mode,
      1,
      () => '?',
    );
    params.push(...keyValues);
    whereSql = whereSql ? `${whereSql} AND ${predicate}` : `WHERE ${predicate}`;
  }

  const limit = safeInt(req.pageSize + 1, 'page size'); // D24's +1 probe
  const offsetSql =
    req.cursor.mode === 'offset' ? ` OFFSET ${safeInt(req.cursor.offset, 'offset')}` : '';

  const sql = [
    `SELECT ${selectList}`,
    `FROM ${relationSql}`,
    whereSql,
    orderBySql ? `ORDER BY ${orderBySql}` : '',
    `LIMIT ${limit}${offsetSql}`,
  ]
    .filter(Boolean)
    .join('\n');

  // Never `prepared` here: mariadb.js's binary/prepared protocol (conn.execute) combined with
  // the textMode typeCast callback corrupts row data (buffer/field misalignment), confirmed with
  // real bound keyset params, not just the zero-param case. conn.query() already binds `?`
  // placeholders client-side (its own escaping, not string interpolation), so keyset params are
  // still safely bound over the text protocol.
  const rawRows = await runQuery<(string | null)[]>(conn, sql, params, ctx, track, {
    rowsAsArray: true,
    textMode: true,
    logParams: true,
  });

  const probedExtra = rawRows.length > req.pageSize;
  const keptRows = probedExtra ? rawRows.slice(0, req.pageSize) : rawRows;

  const builder = createTabularPageBuilder(columns);
  for (const row of keptRows) {
    builder.appendRow(row.slice(0, projectedColumns.length));
  }
  if (reverseRows) builder.reverse();

  const displayRows = reverseRows ? [...keptRows].reverse() : keptRows;

  const position = buildKeysetPosition({
    cursor: req.cursor,
    pageSize: req.pageSize,
    displayRows,
    probedExtra,
    order,
    keysetColumnIndex,
    fingerprint,
    cellAt: (row, i) => row[i],
  });

  return builder.finish(position);
}

export async function countRows(
  conn: Connection,
  ctx: OpCtx,
  track: TrackQuery,
  target: Pick<ReadTarget, 'qualifiedName'>,
  filter: string | null,
): Promise<{ value: number; exact: boolean }> {
  const relationSql = `${quoteIdent(target.qualifiedName.database)}.${quoteIdent(target.qualifiedName.table)}`;
  const sql = [`SELECT count(*) AS n`, `FROM ${relationSql}`, whereClause(filter)]
    .filter(Boolean)
    .join('\n');

  const rows = await runQuery<[string]>(conn, sql, [], ctx, track, {
    rowsAsArray: true,
    textMode: true,
  });
  return { value: parseCountValue(rows[0]?.[0]), exact: true };
}

File: shell/internal/adapters/mysqlfamily/read.go
package mysqlfamily

import (
	"context"
	"database/sql"
	"encoding/hex"
	"regexp"
	"strings"

	"github.com/kirathecat/kira-studio/shell/internal/adapters"
	"github.com/kirathecat/kira-studio/shell/internal/page"
	"github.com/kirathecat/kira-studio/shell/internal/storage/model"
)

// quoteIdent is read.ts's quoteIdent.
func quoteIdent(name string) string {
	if strings.ContainsRune(name, '\x00') {
		panic(adapters.New(adapters.CodeQuery, "identifier contains a NUL byte", nil))
	}
	return "`" + strings.ReplaceAll(name, "`", "``") + "`"
}

// binaryDatabaseTypes is B3's set: go-sql-driver's own DatabaseTypeName() already tells binary
// from text apart by the column's collation (fields.go's typeDatabaseName) — this package never
// needs to consult a collation table itself the way query.ts's typeCastString had to.
var binaryDatabaseTypes = map[string]bool{
	"BLOB": true, "TINYBLOB": true, "MEDIUMBLOB": true, "LONGBLOB": true,
	"BINARY": true, "VARBINARY": true, "GEOMETRY": true, "BIT": true,
}

// cellText is B3's cellText: a binary column's raw bytes render as 0x<hex> (D3, mirroring
// Postgres's bytea handling); everything else is the server's own text bytes, unmodified.
func cellText(raw []byte, dbType string) string {
	if binaryDatabaseTypes[dbType] {
		return "0x" + hex.EncodeToString(raw)
	}
	return string(raw)
}

var tinyint1 = regexp.MustCompile(`^tinyint\(1\)`)
var numberType = regexp.MustCompile(`^(tinyint|smallint|mediumint|int|integer|bigint|decimal|numeric|float|double|bit)\b`)
var temporalType = regexp.MustCompile(`^(date|datetime|timestamp|time|year)\b`)
var binaryType = regexp.MustCompile(`^(binary|varbinary|tinyblob|blob|mediumblob|longblob|geometry)\b`)

// typeClassFor is read.ts's typeClassFor — §5d's MariaDB/MySQL mapping. tinyint(1) is checked
// ahead of the general number match — it is how this family spells boolean.
func typeClassFor(dataType string) page.TypeClass {
	base := strings.ToLower(dataType)
	if tinyint1.MatchString(base) {
		return page.TypeClassBoolean
	}
	if numberType.MatchString(base) {
		return page.TypeClassNumber
	}
	if temporalType.MatchString(base) {
		return page.TypeClassTemporal
	}
	if strings.HasPrefix(base, "json") {
		return page.TypeClassJSON
	}
	if binaryType.MatchString(base) {
		return page.TypeClassBinary
	}
	return page.TypeClassText
}

// readReq is adapter.ts's ReadRequest minus Path.
type readReq struct {
	Projection []string
	Filter     *string
	Sort       *model.SortSpec
	PageSize   int
	Cursor     model.PageCursor
}

func questionPlaceholder(int) string { return "?" }

// readPage is read.ts's readPage — the same eleven-step shape as postgres/read.go's.
func readPage(ctx context.Context, conn *sql.Conn, threadID uint32, op *adapters.OpCtx, track TrackQuery, target ReadTarget, req readReq) (page.TabularPage, error) {
	projectedColumns, err := adapters.ResolveProjection(target.Columns, req.Projection)
	if err != nil {
		return page.TabularPage{}, err
	}
	var tiebreaker []string
	if target.PrimaryKey != nil {
		tiebreaker = target.PrimaryKey
	} else if len(target.UniqueKeys) > 0 {
		tiebreaker = target.UniqueKeys[0]
	}
	order, err := adapters.ComputeEffectiveOrder(req.Sort, target.Columns, tiebreaker)
	if err != nil {
		return page.TabularPage{}, err
	}
	isTextSort := req.Sort != nil && req.Sort.Kind == "text"
	wantsKeyset := req.Cursor.Mode == "after" || req.Cursor.Mode == "before"
	if err := adapters.AssertKeysetSupported(wantsKeyset, isTextSort, order.KeysetEligible); err != nil {
		return page.TabularPage{}, err
	}

	fetch, err := adapters.ResolveFetchColumns(projectedColumns, target.Columns, order, nil)
	if err != nil {
		return page.TabularPage{}, err
	}

	columns := make([]page.ColumnDescriptor, len(projectedColumns))
	for i, c := range projectedColumns {
		columns[i] = page.ColumnDescriptor{
			Name: c.Name, DataType: c.DataType, TypeClass: typeClassFor(c.DataType),
			Nullable: c.Nullable, IsPrimaryKey: c.IsPrimaryKey,
			// P36 D28: not detected here yet — false rather than a guess.
			Generated: false,
		}
	}

	relationSQL := quoteIdent(target.QualifiedName.Database) + "." + quoteIdent(target.QualifiedName.Table)
	selectNames := make([]string, len(fetch.Columns))
	for i, c := range fetch.Columns {
		selectNames[i] = quoteIdent(c.Name)
	}
	selectList := strings.Join(selectNames, ", ")

	var params []any
	whereSQL := adapters.WhereClause(req.Filter)

	fingerprint := adapters.RequestFingerprint(struct {
		Path       QualifiedName   `json:"path"`
		Projection []string        `json:"projection"`
		Filter     *string         `json:"filter"`
		Sort       *model.SortSpec `json:"sort"`
		PageSize   int             `json:"pageSize"`
	}{target.QualifiedName, req.Projection, req.Filter, req.Sort, req.PageSize})

	reverseRows := req.Cursor.Mode == "before" && order.KeysetEligible
	orderBySQL := adapters.BuildScanOrderBy(req.Sort, order, reverseRows, quoteIdent)

	if wantsKeyset && req.Cursor.Mode != "offset" {
		keyValues, err := adapters.DecodePageToken(req.Cursor.Token, fingerprint)
		if err != nil {
			return page.TabularPage{}, err
		}
		if len(keyValues) != len(order.KeysetColumns) {
			return page.TabularPage{}, adapters.New(adapters.CodeQuery, "page token key length does not match the sort key", nil)
		}
		for _, v := range keyValues {
			params = append(params, v)
		}
		quotedKeyColumns := make([]string, len(order.KeysetColumns))
		for i, c := range order.KeysetColumns {
			quotedKeyColumns[i] = quoteIdent(c)
		}
		predicate := adapters.BuildKeysetPredicate(quotedKeyColumns, order.KeysetDirection, req.Cursor.Mode, 1, questionPlaceholder)
		if whereSQL != "" {
			whereSQL += " AND " + predicate
		} else {
			whereSQL = "WHERE " + predicate
		}
	}

	// D24: fetch pageSize + 1 to compute hasMore without a count. Bound before the OFFSET
	// placeholder — params must line up with the two "?"s in "LIMIT ? OFFSET ?" left to right, the
	// same order the SQL text below actually emits them in.
	params = append(params, req.PageSize+1)
	offsetSQL := ""
	if req.Cursor.Mode == "offset" {
		params = append(params, req.Cursor.Offset)
		offsetSQL = " OFFSET ?"
	}

	sqlParts := []string{"SELECT " + selectList, "FROM " + relationSQL}
	if whereSQL != "" {
		sqlParts = append(sqlParts, whereSQL)
	}
	if orderBySQL != "" {
		sqlParts = append(sqlParts, "ORDER BY "+orderBySQL)
	}
	sqlParts = append(sqlParts, "LIMIT ?"+offsetSQL)
	query := strings.Join(sqlParts, "\n")

	rawRows, err := runArrayQuery(ctx, conn, threadID, query, params, op, track, QueryOptions{LogParams: true})
	if err != nil {
		return page.TabularPage{}, err
	}

	probedExtra := len(rawRows) > req.PageSize
	keptRows := rawRows
	if probedExtra {
		keptRows = rawRows[:req.PageSize]
	}

	builder := page.NewTabularPageBuilder(columns)
	for _, row := range keptRows {
		if err := builder.AppendRow(row[:len(projectedColumns)]); err != nil {
			return page.TabularPage{}, err
		}
	}
	if reverseRows {
		builder.Reverse()
	}

	displayRows := keptRows
	if reverseRows {
		displayRows = make([][]*string, len(keptRows))
		for i, row := range keptRows {
			displayRows[len(keptRows)-1-i] = row
		}
	}

	position, err := adapters.BuildKeysetPosition(adapters.KeysetPositionArgs{
		Cursor: req.Cursor, PageSize: req.PageSize, DisplayRowCount: len(displayRows),
		ProbedExtra: probedExtra, Order: order, KeysetColumnIdx: fetch.KeysetColumnIdx,
		Fingerprint: fingerprint,
		CellAt:      func(row, col int) *string { return displayRows[row][col] },
	})
	if err != nil {
		return page.TabularPage{}, err
	}

	return builder.Finish(position), nil
}

// countRows is read.ts's countRows.
func countRows(ctx context.Context, conn *sql.Conn, threadID uint32, op *adapters.OpCtx, track TrackQuery, target QualifiedName, filter *string) (adapters.CountResult, error) {
	relationSQL := quoteIdent(target.Database) + "." + quoteIdent(target.Table)
	sqlParts := []string{"SELECT count(*) AS n", "FROM " + relationSQL}
	if where := adapters.WhereClause(filter); where != "" {
		sqlParts = append(sqlParts, where)
	}
	query := strings.Join(sqlParts, "\n")

	rows, err := runArrayQuery(ctx, conn, threadID, query, nil, op, track, QueryOptions{})
	if err != nil {
		return adapters.CountResult{}, err
	}
	var raw any
	if len(rows) > 0 && len(rows[0]) > 0 && rows[0][0] != nil {
		raw = *rows[0][0]
	}
	value, err := adapters.ParseCountValue(raw)
	if err != nil {
		return adapters.CountResult{}, err
	}
	return adapters.CountResult{Value: value, Exact: true}, nil
}
